src/Raven.Server/Documents/ETL/EtlProcess.cs

```diff
@@ -0,0 +1,220 @@
+            var tombstones = Database.DocumentsStorage.GetTombstonesFrom(context, _configuration.Collection, Statistics.LastProcessedEtag + 1, 0, int.MaxValue);
+
+            using (var documentsIt = documents.GetEnumerator())
+            using (var tombstonesIt = tombstones.GetEnumerator())
+            {
+                using (var merged = new MergedEnumerator<TExtracted>())
+                {
+                    merged.AddEnumerator(ConvertDocsEnumerator(documentsIt));
+                    merged.AddEnumerator(ConvertTombstonesEnumerator(tombstonesIt));
+
+                    while (merged.MoveNext())
+                    {
+                        yield return merged.Current;
+                    }
+                }
+            }
+        }
+
+        public abstract IEnumerable<TTransformed> Transform(IEnumerable<TExtracted> items, DocumentsOperationContext context);
+
+        public abstract bool Load(IEnumerable<TTransformed> items);
+
+        public abstract bool CanContinueBatch();
+
+        protected abstract void LoadLastProcessedEtag(DocumentsOperationContext context);
+
+        protected abstract void StoreLastProcessedEtag(DocumentsOperationContext context);
+
+        public void NotifyAboutWork()
+        {
+            _waitForChanges.Set();
+        }
+
+        public void Start()
+        {
+            if (_thread != null)
+                return;
+
+            if (_configuration.Disabled)
+                return;
+
+            _thread = new Thread(() =>
+            {
+                // This has lower priority than request processing, so we let the OS
+                // schedule this appropriately
+                Threading.TrySettingCurrentThreadPriority(ThreadPriority.BelowNormal);
+
+                Run();
+            })
+            {
+                Name = $"{Tag} process: {Name}",
+                IsBackground = true
+            };
+
+            if (Logger.IsInfoEnabled)
+                Logger.Info($"Starting {Tag} process: '{Name}'.");
+
+            _thread.Start();
+        }
+
+        public void Stop()
+        {
+            if (_thread == null)
+                return;
+
+            if (Logger.IsInfoEnabled)
+                Logger.Info($"Stopping {Tag} process: '{Name}'.");
+
+            _cts.Cancel();
+
+            var thread = _thread;
+            _thread = null;
+
+            if (Thread.CurrentThread != thread) // prevent a deadlock
+                thread.Join();
+        }
+
+        public void Run()
+        {
+            while (CancellationToken.IsCancellationRequested == false)
+            {
+                _waitForChanges.Reset();
+
+                var startTime = Database.Time.GetUtcNow();
+                var duration = Stopwatch.StartNew();
+
+
+                // TODO arek
+                //if (Statistics.SuspendUntil.HasValue && Statistics.SuspendUntil.Value > Database.Time.GetUtcNow())
```

Comment:
It's was a part of SQL replication. `SuspendUntil` data is set here as `newTime`:

https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Documents/ETL/Providers/SQL/SqlEtl.cs#L108-L119

I think I will handle that differently and make it specific only for SQL ETL instead of having it in a base `EtlProcess` class.